File: 0014_Longest_Common_Prefix/0014_Longest_Common_Prefix.py
"""
This is the horizontal comparison version
Write a function to find the longest common prefix string amongst an array of strings.
"""

class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        if len(strs) == 0:
            return ""
        prefix = strs[0]
        for string in strs[1:]:
            while string[:len(prefix)] != prefix and prefix:
                prefix = prefix[:len(prefix) - 1]
                if not prefix:
                    return ""
        return prefix

"""
Example 1:
Input: strs = ["flower","flow","flight"]
Output: "fl"

Example 2:
Input: strs = ["dog","racecar","car"]
Output: ""

Time: O(n) (n is the sum of all chars in all strings)
"""

File: 0014_Longest_Common_Prefix/0014_Longest_Common_Prefix.cc
/*
This is the vertical comparison version
Write a function to find the longest common prefix string amongst an array of strings.
*/

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.size() == 0) return "";
        for (int i = 0; i < strs.at(0).length(); i++) {
            char cur_char = strs.at(0)[i];
            for (int j = 1; j < strs.size(); j++) {
                if (strs.at(j)[i] != cur_char || i > strs.at(j).length() - 1) {
                    return strs.at(0).substr(0, i);
                }
            }
        }
        return strs.at(0);
    }
};

/*
Example 1:
Input: strs = ["flower","flow","flight"]
Output: "fl"

Example 2:
Input: strs = ["dog","racecar","car"]
Output: ""

Time: O(n) (n is the sum of all chars in all strings)
*/
